research-app/src/App.vue

```diff
@@ -1213,6 +1361,369 @@ export default class App extends WWTAwareComponent {
+
+      const nextTypesPresent = nextTypes.filter(t => messageTypes.indexOf(t) >= 0);
+
+      if (completionInfo) {
+        const [completedType, completedMatcher] = completionInfo;
+        const handler: (msg: any) => boolean = (msg) => {
+
+          //@ts-ignore
+          const matchingMessages = messagesOfType.filter(x => completedMatcher(x, msg));
+          matchingMessages.forEach(m => {
+            messagesOfType.splice(messagesOfType.indexOf(m), 1);
+          });
+          if (messagesOfType.length === 0) {
+            this.messageHandlers.delete(msgType);
+            finishedMessageTypes.push(msgType);
+            nextTypesPresent.filter(prerequisitesMet).forEach(addMessagesOfType);
+          }
+          return true;
+        };
+        this.messageHandlers.set(completedType, handler);
+      } else {
+        finishedMessageTypes.push(msgType);
+        nextTypesPresent.filter(prerequisitesMet).forEach(addMessagesOfType);
+      }
+    }
+
+    rootTypes.forEach(addMessagesOfType);
+  }
+
+  messageQueue: Message[] = [];
+
+  @Watch("messageQueue")
+  handleMessageQueueUpdate(queue: Message[]): void {
+    while (queue.length > 0) {
+      const message = queue.shift();
+      this.onMessage(message);
+    }
+  }
+
+  stateAsUrl(): string {
+
+    const coordinatesMessage: classicPywwt.CenterOnCoordinatesMessage = {
+      event: "center_on_coordinates",
+      ra: this.wwtRARad * R2D,
+      dec: this.wwtDecRad * R2D,
+      fov: this.wwtZoomDeg / 6,
+      instant: true,
+    };
+
+    let backgroundMessage: classicPywwt.SetBackgroundByNameMessage | null = null;
+    if (this.wwtBackgroundImageset !== null) {
+      let name = this.wwtBackgroundImageset.get_name();
+      if (name === 'DSS') {
+        name = 'Digitized Sky Survey (Color)'
+      };
+      backgroundMessage = {
+        event: "set_background_by_name",
+        name: name,
+      };
+    }
+    
+
+    let foregroundMessage: classicPywwt.SetForegroundByNameMessage | null = null;
+    if (this.wwtForegroundImageset !== null) {
+      foregroundMessage = {
+        event: "set_foreground_by_name",
+        name: this.wwtForegroundImageset.get_name(),
+      };
+    }
+
+    const threadId = Guid.create();
+    const catalogs = this.hipsCatalogs();
+    const loadCatalogsMessages: layers.LoadHipsCatalogMessage[] = [];
+    const catalogSettingsMessages: layers.MultiModifyTableLayerMessage[] = [];
+    catalogs.forEach(catalog => {
+      const layer = this.layerForHipsCatalog(catalog.name);
+      if (layer !== null) {
+        const id = layer.id.toString();
+        loadCatalogsMessages.push({
+          event: "layer_hipscat_load",
+          threadId: threadId, // We need this to get a response
+          tableId: id,
+          name: catalog.name,
+        });
+
+
+        const state: SpreadSheetLayerState = this.wwtSpreadSheetLayers[id];
+        const values: any[] = spreadSheetLayerSettingNames.map(setting => (state as any)["get_" + setting]());
```

Comment:
This kind of functionality is centralized in functions like [`@wwtelescope/engine-helpers::extractSpreadSheetLayerSettings`][x]. Can they be used here?

[x]: https://docs.worldwidetelescope.org/webgl-reference/latest/apiref/engine-helpers/modules.html#extractspreadsheetlayersettings